lib/proxy/client.go

```diff
@@ -0,0 +1,505 @@
+			}
+		}
+	}
+}
+
+func (c *Client) updateConnections(proxies []types.Server) error {
+	c.Lock()
+	defer c.Unlock()
+	var errs []error
+
+	toDial := make(map[string]types.Server)
+	for _, proxy := range proxies {
+		toDial[proxy.GetName()] = proxy
+	}
+
+	for id, conn := range c.conns {
+		_, ok := toDial[id]
+
+		// delete nonexistent connections
+		if !ok {
+			c.closeConn(conn)
+			continue
+		}
+
+		// test existing connections
+		if err := c.testConnection(conn); err != nil {
+			errs = append(errs, err)
+			c.closeConn(conn)
+			continue
+		}
+	}
+
+	for id, proxy := range toDial {
+		// skips itself
+		if id == c.config.ID {
+			continue
+		}
+
+		// skip existing connection. they've been tested above.
+		if _, ok := c.conns[id]; ok {
+			continue
+		}
+
+		// establish new connections
+		conn, err := c.connect(id, proxy.GetPeerAddr())
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if err := c.testConnection(conn); err != nil {
+			errs = append(errs, err)
+			c.closeConn(conn)
+			continue
+		}
+
+		c.conns[id] = conn
+	}
+
+	return trace.NewAggregate(errs...)
+}
+
+func (c *Client) cleanup() error {
+	c.RLock()
+	var errs []error
+	toDelete := make(map[string]struct{})
+	for id, conn := range c.conns {
+		err := c.testConnection(conn)
+		if err == nil {
+			continue
+		}
+		errs = append(errs, err)
+		toDelete[id] = struct{}{}
+	}
+
+	c.RUnlock()
+	if len(toDelete) == 0 {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+	for id := range toDelete {
+		c.closeConn(c.conns[id])
+	}
+
+	return trace.NewAggregate(errs...)
+}
+
+// Dial dials a node through a peer proxy.
+func (c *Client) Dial(
+	proxyIDs []string,
+	nodeID string,
+	src net.Addr,
+	dst net.Addr,
+	tunnelType types.TunnelType,
+) (net.Conn, error) {
+	stream, _, err := c.dial(proxyIDs)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// send dial request as the first frame
+	if err = stream.Send(&clientapi.Frame{
+		Message: &clientapi.Frame_DialRequest{
+			DialRequest: &clientapi.DialRequest{
+				NodeID:     nodeID,
+				TunnelType: tunnelType,
+				Source: &clientapi.NetAddr{
+					Addr:    src.String(),
+					Network: src.Network(),
+				},
+				Destination: &clientapi.NetAddr{
+					Addr:    dst.String(),
+					Network: dst.Network(),
+				},
+			},
+		},
+	}); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	conn := newStreamConn(stream, src, dst)
+	go conn.start()
+
+	return conn, nil
+}
+
+// Close closes all existing client connections.
+func (c *Client) Close() error {
+	c.Lock()
+	defer c.Unlock()
+
+	var errs []error
+	for _, conn := range c.conns {
+		if err := c.closeConn(conn); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	c.cancel()
+	return trace.NewAggregate(errs...)
+}
+
+// closeConn closes and removes a clientConn from cache.
+// This function is not thread safe.
+func (c *Client) closeConn(conn *clientConn) (err error) {
+	conn.cancel()
+	conn.wg.Wait() // wait for streams to gracefully end
```

Comment:
`closeConn` is called from a few different places above while holding the global lock could this block for a while? Maybe we can defer graceful shutdowns to be handled by the reverse tunnel server or where ever this client library is used.

Reply:
I've only tested this on a small scale and while writing tests, so don't have a good answer about performance or holding the lock for an extended period of time, but I am going to investigate and try to get some numbers or stress test that somehow.
But the issue mainly comes from the buffering grpc does in the background. I just wanted to be able to test cases such as:
- closed client connections -> streams should error out and not accept any more send messages
- streams didn't always receive eof errors a connection is dropped, so we couldn't log this properly in prometheus.
...

that's mainly why I elected to closestreams manually before finalizing closing the connection. I am not sure it does affect things in real life, but I found lots of complaints about that on the grpc-go git repo, mostly people asking for graceful shutdown on the client side just like there is one on the server side